### PR TITLE
Fix parsing for dot following arr and [*] path and add `TABLE` keyword

### PR DESCRIPTION
**Pull request by am357**
*Issue #, if available:* #137 #108 

*Description of changes:*
- Fix parsing for path whildcard and array with a dot
    
    This change fixes parsing of expressions like the following:
    `foo(x, y)[*].*.b[5]`
    
    In the above expression a `.` follows a path whildcard

- Add `TABLE` keyword
    This is to ensure the following syntax conformance test passes:
    
```rust
    fn invalid_path_component_keyword_in_path_test() {
        let statement = r#"SELECT foo.id, foo.table FROM `[{id: 1, table: "foos"}]` AS foo"#;
        let res = partiql_parser::Parser::default().parse(statement);
        assert!(
            res.is_err(),
            "For `{}`, expected `Err(_)`, but was `{:#?}`",
            statement,
            res
        );
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-parser/src/parse/partiql.lalrpop [hunk 1]
```diff
@@ -1294,6 +1324,7 @@ extern {
         "PRESERVE" => lexer::Token::Preserve,
         "RIGHT" => lexer::Token::Right,
         "SELECT" => lexer::Token::Select,
+        "Table" => lexer::Token::Table,
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:1327]: s/Table/TABLE/

**Review by jpschorr — APPROVED** [review]
Some cleanup recommendations; otherwise LGTM

File: partiql-parser/src/parse/partiql.lalrpop [hunk 2]
```diff
@@ -954,15 +973,13 @@ PathSteps: Vec<ast::PathStep> = {
         let mut steps = path;
         steps.push(step);
         steps
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:975]:
```suggestion
vec![ast::PathStep::PathExpr( ast::PathExpr{ index: Box::new(v) })]
```

File: partiql-parser/src/parse/partiql.lalrpop [hunk 3]
```diff
@@ -954,15 +973,13 @@ PathSteps: Vec<ast::PathStep> = {
...
-        // ast::Path{ root:path.root, steps }
     },
-    <lo:@L> <path:PathSteps>  "[" "*" "]" <hi:@R> => {
+    <lo:@L> <path:PathSteps> "[" "*" "]" <hi:@R> => {
          let step = ast::PathStep::PathWildCard;
 
          let mut steps = path;
          steps.push(step);
          steps
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:982]:
```suggestion
vec![ast::PathStep::PathWildCard];
```

File: partiql-parser/src/parse/partiql.lalrpop [hunk 4]
```diff
@@ -983,7 +1000,21 @@ PathSteps: Vec<ast::PathStep> = {
...
+    "[" "*" "]" => {
+        let steps = vec![ast::PathStep::PathWildCard];
+        steps
+    },
+    "[" <expr:ExprQuery> "]" => {
+        let steps = vec![ast::PathStep::PathExpr(
+            ast::PathExpr{
+            index: Box::new(*expr),
+        })];
+        steps
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:1013]:
```suggestion
vec![ast::PathStep::PathExpr( ast::PathExpr{ index: Box::new(*expr) })]
```

File: partiql-parser/src/parse/partiql.lalrpop [hunk 5]
```diff
@@ -983,7 +1000,21 @@ PathSteps: Vec<ast::PathStep> = {
         let mut steps = path;
         steps.push(step);
         steps
-        // ast::Path{ root:path.root, steps }
+    },
+    "[" "*" "]" => {
+        let steps = vec![ast::PathStep::PathWildCard];
+        steps
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:1006]:
```suggestion
vec![ast::PathStep::PathWildCard];
```

File: partiql-parser/src/parse/partiql.lalrpop [hunk 6]
```diff
@@ -983,7 +1000,21 @@ PathSteps: Vec<ast::PathStep> = {
...
+    "[" <expr:ExprQuery> "]" => {
+        let steps = vec![ast::PathStep::PathExpr(
+            ast::PathExpr{
+            index: Box::new(*expr),
+        })];
+        steps
+    },
+    "*" => {
+        let steps = vec![ast::PathStep::PathUnpivot];
+        steps
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:1017]:
```suggestion
vec![ast::PathStep::PathUnpivot]
```

File: partiql-parser/src/parse/partiql.lalrpop [hunk 7]
```diff
@@ -992,7 +1023,6 @@ PathSteps: Vec<ast::PathStep> = {
             });
         let steps = vec![step];
         steps
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:1025]:
```suggestion
vec![ast::PathStep::PathExpr( ast::PathExpr{ index: Box::new(*expr) })]
```

File: partiql-parser/src/parse/partiql.lalrpop (hunk 1 above)
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:1327]:
```suggestion
        "TABLE" => lexer::Token::Table,
```

File: partiql-parser/src/parse/partiql.lalrpop [hunk 8]
```diff
@@ -750,11 +750,37 @@ ExprPrecedence03: ast::Expr = {
...
-    <l:ExprPrecedence01> "." "*" => ast::Path {
-        root:Box::new(l), steps:vec![ast::PathStep::PathUnpivot]
-     },
+    <l:ExprPrecedence01> "[" "*" "]" "." <s:PathSteps> => {
+        let mut steps = vec![ast::PathStep::PathWildCard];
+        steps.extend(s);
+        ast::Path {
+            root:Box::new(l),
+            steps
+        }
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:768]:
```suggestion
        let step = ast::PathStep::PathWildCard;
        ast::Path {
            root: Box::new(l),
            steps: std::iter::once(step).chain(s.into_iter()).collect()
        }
```

File: partiql-parser/src/parse/partiql.lalrpop [hunk 9]
```diff
@@ -750,11 +750,37 @@ ExprPrecedence03: ast::Expr = {
...
+                index: Box::new(*expr),
+            }
+        );
+
+        let mut steps = vec![step];
+        steps.extend(s);
+        ast::Path {
+            root:Box::new(l),
+            steps
+        }
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:782]:
```suggestion
        ast::Path {
            root: Box::new(l),
            steps: std::iter::once(step).chain(s.into_iter()).collect()
        }
```

**Review by jpschorr — APPROVED** [review]

**Review by alancai98 — APPROVED** [review]

**Pull request merged**